applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.cpp

```diff
@@ -0,0 +1,334 @@
+        {
+           const IndexType coord_dir = DifferentiationUtility::GetCoordinateDirection(rDesignVariable);
+
+            // define working variables
+            Vector RHS_unperturbed;
+            Vector RHS_perturbed;
+            ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+            // compute RHS before perturbion
+            rElement.CalculateRightHandSide(RHS_unperturbed, copy_process_info);
+
+            if ( rOutput.size() != RHS_unperturbed.size() )
+                rOutput.resize(RHS_unperturbed.size(), false);
+
+            // perturb the design variable
+            rNode.GetInitialPosition()[coord_dir] += rPertubationSize;
+            rNode.Coordinates()[coord_dir] += rPertubationSize;
+
+            // compute LHS after perturbation
+            rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+
+            //compute derivative of RHS w.r.t. design variable with finite differences
+            noalias(rOutput) = (RHS_perturbed - RHS_unperturbed) / rPertubationSize;
+
+             // unperturb the design variable
+            rNode.GetInitialPosition()[coord_dir] -= rPertubationSize;
+            rNode.Coordinates()[coord_dir] -= rPertubationSize;
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+        }
+        else
+        {
+            KRATOS_WARNING("DifferentiationUtility") << "Unsupported nodal design variable: " << rDesignVariable << std::endl;
+            if ( (rOutput.size() != 0) )
+                rOutput.resize(0,false);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void DifferentiationUtility::CalculateLeftHandSideDerivative(Element& rElement,
+                                                const array_1d_component_type& rDesignVariable,
+                                                Node<3>& rNode,
+                                                const double& rPertubationSize,
+                                                Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        #ifdef _OPENMP
+            KRATOS_ERROR_IF(omp_get_thread_num() > 0) <<
+                "DifferentiationUtility::CalculateLeftHandSideDerivative " <<
+                "is not thread safe for shape derivatives!" << omp_get_thread_num();
+        #endif
+
+        if( rDesignVariable == SHAPE_X || rDesignVariable == SHAPE_Y || rDesignVariable == SHAPE_Z )
+        {
+            const IndexType coord_dir = DifferentiationUtility::GetCoordinateDirection(rDesignVariable);
+
+            // define working variables
+            Matrix LHS_unperturbed;
+            Matrix LHS_perturbed;
+            Vector dummy;
+            ProcessInfo copy_process_info = rCurrentProcessInfo;
```

Comment:
why are we making a copy of the process info?

Reply:
This is due to the fact that the Process Info is passed as ```const```. This is oriented on ```CalculateSensitivityMatrix```. I think it is the best to skip this constant passing.

Reply:
Now i remember. If we need it non const in this function, we should also show that in the parameters list of this function and pass it non const.

Reply:
I will pass the process info as non const.